diesel/src/connection/statement_cache.rs

```diff
@@ -36,30 +37,58 @@ impl<DB, Statement> StatementCache<DB, Statement> where
         source: &T,
         bind_types: &[DB::TypeMetadata],
         prepare_fn: F,
-    ) -> QueryResult<Statement> where
+    ) -> QueryResult<MaybeCached<Statement>> where
         T: QueryFragment<DB> + QueryId,
         F: FnOnce(&str) -> QueryResult<Statement>,
-        Statement: Clone,
     {
         use std::collections::hash_map::Entry::{Occupied, Vacant};
 
         let cache_key = try!(StatementCacheKey::for_source(source, bind_types));
-        let mut cache = self.cache.borrow_mut();
-
-        match cache.entry(cache_key) {
-            Occupied(entry) => Ok(entry.get().clone()),
-            Vacant(entry) => {
-                let statement = {
-                    let sql = try!(entry.key().sql(source));
-                    prepare_fn(&sql)
-                };
-
-                if !source.is_safe_to_cache_prepared() {
-                    return statement;
-                }
 
-                Ok(entry.insert(try!(statement)).clone())
+        if !source.is_safe_to_cache_prepared() {
+            let sql = try!(cache_key.sql(source));
+            return prepare_fn(&sql).map(MaybeCached::CannotCache)
+        }
+
+        refmut_map_result(self.cache.borrow_mut(), |cache| {
+            match cache.entry(cache_key) {
+                Occupied(entry) => Ok(entry.into_mut()),
+                Vacant(entry) => {
+                    let statement = {
+                        let sql = try!(entry.key().sql(source));
+                        prepare_fn(&sql)
+                    };
+
+                    Ok(entry.insert(try!(statement)))
+                }
             }
+        }).map(MaybeCached::Cached)
+    }
+}
+
+#[doc(hidden)]
+#[allow(missing_debug_implementations)]
+pub enum MaybeCached<'a, T: 'a> {
+    CannotCache(T),
+    Cached(RefMut<'a, T>),
+}
+
+impl<'a, T> Deref for MaybeCached<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        match *self {
+            MaybeCached::CannotCache(ref x) => x,
+            MaybeCached::Cached(ref x) => &**x,
```

Comment:
This is how I program in C, too! Add `&` and `*` until it works :trollface: 

FTR, this deref's the RefMut, then its content, and returns a reference to that.